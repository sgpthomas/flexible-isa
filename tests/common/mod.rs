use std::{
    fmt::Display,
    path::{Path, PathBuf},
};

use flexible_isa::{cli, run, HalideExprOp, InstructionSelect};
use itertools::Itertools;

const BENCHMARKS_PATH: &str = "cache";
#[allow(unused)]
const BENCHMAKRS: &[&str] = &[
    "add",
    "average_pool",
    "blur3x3",
    "camera_pipe",
    "conv3x3a16",
    "conv3x3a32",
    "conv_nn",
    "depthwise_conv",
    "dilate3x3",
    // "elementwise",
    // "fully_connected",
    "gaussian3x3",
    "gaussian5x5",
    "gaussian7x7",
    "l2norm",
    "matmul",
    "max_pool",
    "mean",
    "median3x3",
    "mul",
    "sobel3x3",
    "softmax",
];

pub fn test_single_isa<S: AsRef<Path> + Display>(name: S) -> anyhow::Result<()> {
    let mut path = PathBuf::from(BENCHMARKS_PATH).join(&name);
    path.set_extension("stmt");

    let args: cli::Args = cli::Args::new(&[path]).learn(true);
    let isa = run(args)?;

    let instr_hist = InstructionSelect::from_recexpr(&isa.expressions);
    let hist: Vec<_> = instr_hist
        .iter()
        .filter_map(|(op, count)| {
            if let HalideExprOp::Instruction(i) = op {
                Some((i, count))
            } else {
                None
            }
        })
        // sort first by count, then by index
        .sorted_by_key(|(i, count)| (*count, *i))
        .map(|(i, count)| (i, count, isa.instructions[i].to_string()))
        .collect();

    insta::with_settings!({
        description => format!("generate isa for {name}"),
        omit_expression => true,
        snapshot_suffix => format!("{name}")
    }, {
        insta::assert_debug_snapshot!(hist)
    });

    Ok(())
}
