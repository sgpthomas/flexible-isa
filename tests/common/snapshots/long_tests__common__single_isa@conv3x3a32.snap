---
source: tests/common/mod.rs
description: generate isa for conv3x3a32
---
[
    (
        6,
        1,
        "(+ (cast<int32> (- (cast<int32> ?x0) (cast<int32> ?x1))) (cast<int32> ?x2))",
    ),
    (
        7,
        1,
        "(+ (cast<int32> (* (cast<int32> (+ (cast<int32> ?#1) (cast<int32> ?x0))) (cast<int32> 4))) (cast<int32> 8))",
    ),
    (
        25,
        1,
        "(+ (cast<int32> (call[max] (cast<int32> (- (cast<int32> (call[min] (cast<int32> (+ (cast<int32> ?x0) (cast<int32> ?x1))) (cast<int32> ?#4))) (cast<int32> ?x2))) (cast<int32> ?x3))) (cast<int32> 1))",
    ),
    (
        27,
        1,
        "(+ (cast<int32> (call[max] (cast<int32> ?x0) (cast<int32> ?x1))) (cast<int32> ?x2))",
    ),
    (
        30,
        1,
        "(+ (cast<int32> ?x0) (cast<int32> 2))",
    ),
    (
        31,
        1,
        "(+ (cast<int32> ?x0) (cast<int32> 3))",
    ),
    (
        35,
        1,
        "(+ (cast<int32> ?x0) (cast<int32> 128))",
    ),
    (
        36,
        1,
        "(+ (cast<int32> ?x0) (cast<int32> (neg (cast<int32> 1))))",
    ),
    (
        39,
        1,
        "(+ (cast<int32x128> (+ (cast<int32x128> (+ (cast<int32x128> (call[widening_add] (cast<int16x128> (call[shift_left] (cast<int16x128> (call[int16x128] (cast<uint8x128> (get ?#15 (cast<int32x128> (call[ramp] (cast<int32> ?x0) (cast<int32> 1) (cast<int32> 128))))))) (cast<uint16x128> (call[x128] (cast<uint16> 2))))) (cast<int16x128> (call[widening_mul] (cast<uint8x128> (get ?#16 (cast<int32x128> (call[ramp] (cast<int32> (+ (cast<int32> (* (cast<int32> ?x1) (cast<int32> 128))) (cast<int32> ?x2))) (cast<int32> 1) (cast<int32> 128))))) (cast<int8x128> (call[x128] (cast<int8> 7))))))) (cast<int32x128> (call[widening_add] (cast<int16x128> (call[widening_mul] (cast<uint8x128> (get ?#17 (cast<int32x128> (call[ramp] (cast<int32> (+ (cast<int32> (* (cast<int32> ?x1) (cast<int32> 128))) (cast<int32> ?x3))) (cast<int32> 1) (cast<int32> 128))))) (cast<int8x128> (call[x128] (cast<int8> 10))))) (cast<int16x128> (call[widening_mul] (cast<uint8x128> (get ?#18 (cast<int32x128> (call[ramp] (cast<int32> (+ (cast<int32> (* (cast<int32> ?x4) (cast<int32> 128))) (cast<int32> ?x5))) (cast<int32> 1) (cast<int32> 128))))) (cast<int8x128> (call[x128] (cast<int8> 5))))))))) (cast<int32x128> (call[widening_add] (cast<int16x128> (call[shift_left] (cast<int16x128> (call[int16x128] (cast<uint8x128> (get ?#19 (cast<int32x128> (call[ramp] (cast<int32> (+ (cast<int32> (* (cast<int32> ?x4) (cast<int32> 128))) (cast<int32> ?x6))) (cast<int32> 1) (cast<int32> 128))))))) (cast<uint16x128> (call[x128] (cast<uint16> 3))))) (cast<int16x128> (call[widening_mul] (cast<uint8x128> (get ?#20 (cast<int32x128> (call[ramp] (cast<int32> (+ (cast<int32> (* (cast<int32> ?x4) (cast<int32> 128))) (cast<int32> ?x7))) (cast<int32> 1) (cast<int32> 128))))) (cast<int8x128> (call[x128] (cast<int8> 11))))))))) (cast<int32x128> (call[widening_add] (cast<int16x128> (call[widening_mul] (cast<uint8x128> (get ?#21 (cast<int32x128> (call[ramp] (cast<int32> (+ (cast<int32> (* (cast<int32> ?x8) (cast<int32> 128))) (cast<int32> ?x9))) (cast<int32> 1) (cast<int32> 128))))) (cast<int8x128> (call[x128] (cast<int8> 6))))) (cast<int16x128> (call[widening_mul] (cast<uint8x128> (get ?#22 (cast<int32x128> (call[ramp] (cast<int32> (+ (cast<int32> (* (cast<int32> ?x8) (cast<int32> 128))) (cast<int32> ?x10))) (cast<int32> 1) (cast<int32> 128))))) (cast<int8x128> (call[x128] (cast<int8> 9))))))))",
    ),
    (
        57,
        1,
        "(- (cast<int32> ?x0) (cast<int32> ?x1))",
    ),
    (
        62,
        1,
        "(* (cast<int32> (+ (cast<int32> ?x0) (cast<int32> 1))) (cast<int32> ?x1))",
    ),
    (
        76,
        1,
        "(== ?#0 ?#1)",
    ),
    (
        77,
        1,
        "(&& ?#0 ?#1)",
    ),
    (
        79,
        1,
        "(get ?x0 (cast<int32x128> (call[ramp] (cast<int32> ?x1) (cast<int32> 1) (cast<int32> 128))))",
    ),
    (
        87,
        1,
        "(get ?#1 (cast<int32x128> (call[ramp] (cast<int32> (+ (cast<int32> (* (cast<int32> ?x0) (cast<int32> 128))) (cast<int32> 258))) (cast<int32> 1) (cast<int32> 128))))",
    ),
    (
        90,
        1,
        "(get ?#1 (cast<int32x128> (call[ramp] (cast<int32> (+ (cast<int32> (* (cast<int32> ?x0) (cast<int32> 128))) (cast<int32> 386))) (cast<int32> 1) (cast<int32> 128))))",
    ),
    (
        93,
        1,
        "(get ?#1 (cast<int32x128> (call[ramp] (cast<int32> (+ (cast<int32> (* (cast<int32> ?x0) (cast<int32> 128))) (cast<int32> 514))) (cast<int32> 1) (cast<int32> 128))))",
    ),
    (
        94,
        1,
        "(get ?#1 (cast<int32x128> (call[ramp] (cast<int32> (+ (cast<int32> (* (cast<int32> ?#2) (cast<int32> 128))) (cast<int32> ?x0))) (cast<int32> 1) (cast<int32> 128))))",
    ),
    (
        2,
        2,
        "(+ (cast<int32> (+ (cast<int32> (* (cast<int32> ?x0) (cast<int32> 128))) (cast<int32> ?x1))) (cast<int32> 128))",
    ),
    (
        8,
        2,
        "(+ (cast<int32> (* (cast<int32> (+ (cast<int32> ?#1) (cast<int32> ?x0))) (cast<int32> 4))) (cast<int32> 13))",
    ),
    (
        10,
        2,
        "(+ (cast<int32> (* (cast<int32> ?x0) (cast<int32> 4))) (cast<int32> ?x1))",
    ),
    (
        29,
        2,
        "(+ (cast<int32> ?x0) (cast<int32> 1))",
    ),
    (
        33,
        2,
        "(+ (cast<int32> ?x0) (cast<int32> 6))",
    ),
    (
        37,
        2,
        "(+ (cast<int32> ?x0) (cast<int32> (neg (cast<int32> ?x1))))",
    ),
    (
        56,
        2,
        "(- (cast<int32> ?x0) (cast<int32> (call[max] (cast<int32> (call[min] (cast<int32> ?x1) (cast<int32> ?x2))) (cast<int32> ?x3))))",
    ),
    (
        60,
        2,
        "(* (cast<int32> (+ (cast<int32> (call[max] (cast<int32> (call[min] (cast<int32> (+ (cast<int32> ?x0) (cast<int32> ?x1))) (cast<int32> ?#2))) (cast<int32> 1))) (cast<int32> (neg (cast<int32> 1))))) (cast<int32> ?#3))",
    ),
    (
        70,
        2,
        "(/ (cast<int32> (+ (cast<int32> ?x0) (cast<int32> ?x1))) (cast<int32> 128))",
    ),
    (
        71,
        2,
        "(/ (cast<int32> (+ (cast<int32> ?x0) (cast<int32> ?x1))) (cast<int32> ?x2))",
    ),
    (
        74,
        2,
        "(<= (cast<int32> (call[max] (cast<int32> ?x0) (cast<int32> 1))) (cast<int32> ?x1))",
    ),
    (
        81,
        2,
        "(get ?#1 (cast<int32x128> (+ (cast<int32x128> (call[max] (cast<int32x128> (call[min] (cast<int32x128> (call[ramp] (cast<int32> (+ (cast<int32> (* (cast<int32> ?#2) (cast<int32> 128))) (cast<int32> (neg (cast<int32> 1))))) (cast<int32> 1) (cast<int32> 128))) (cast<int32x128> (call[x128] (cast<int32> (+ (cast<int32> ?#3) (cast<int32> (neg (cast<int32> 1))))))))) (cast<int32x128> (call[x128] (cast<int32> 0))))) (cast<int32x128> (call[x128] (cast<int32> ?x0))))))",
    ),
    (
        95,
        2,
        "(get ?#2 (cast<int32x128> (+ (cast<int32x128> (call[max] (cast<int32x128> (call[min] (cast<int32x128> (call[ramp] (cast<int32> ?x0) (cast<int32> 1) (cast<int32> 128))) (cast<int32x128> (call[x128] (cast<int32> (+ (cast<int32> ?#3) (cast<int32> (neg (cast<int32> 1))))))))) (cast<int32x128> (call[x128] (cast<int32> 0))))) (cast<int32x128> (call[x128] (cast<int32> ?x1))))))",
    ),
    (
        0,
        3,
        "(neg (cast<int32> ?x0))",
    ),
    (
        28,
        3,
        "(+ (cast<int32> (call[min] (cast<int32> (call[max] (cast<int32> ?x0) (cast<int32> 1))) (cast<int32> (+ (cast<int32> ?x0) (cast<int32> 6))))) (cast<int32> (neg (cast<int32> 1))))",
    ),
    (
        34,
        3,
        "(+ (cast<int32> ?x0) (cast<int32> 127))",
    ),
    (
        40,
        3,
        "(+ (cast<int32x128> (+ (cast<int32x128> (+ (cast<int32x128> (call[widening_add] (cast<int16x128> (call[shift_left] (cast<int16x128> (call[int16x128] (cast<uint8x128> (get ?#16 (cast<int32x128> (call[ramp] (cast<int32> (+ (cast<int32> (* (cast<int32> ?x0) (cast<int32> 128))) (cast<int32> ?x1))) (cast<int32> 1) (cast<int32> 128))))))) (cast<uint16x128> (call[x128] (cast<uint16> 2))))) (cast<int16x128> (call[widening_mul] (cast<uint8x128> (get ?#17 (cast<int32x128> (call[ramp] (cast<int32> (+ (cast<int32> (* (cast<int32> ?x0) (cast<int32> 128))) (cast<int32> ?x2))) (cast<int32> 1) (cast<int32> 128))))) (cast<int8x128> (call[x128] (cast<int8> 7))))))) (cast<int32x128> (call[widening_add] (cast<int16x128> (call[widening_mul] (cast<uint8x128> (get ?#18 (cast<int32x128> (call[ramp] (cast<int32> (+ (cast<int32> (* (cast<int32> ?x0) (cast<int32> 128))) (cast<int32> ?x3))) (cast<int32> 1) (cast<int32> 128))))) (cast<int8x128> (call[x128] (cast<int8> 10))))) (cast<int16x128> (call[widening_mul] (cast<uint8x128> (get ?#19 (cast<int32x128> (call[ramp] (cast<int32> (+ (cast<int32> (* (cast<int32> ?x4) (cast<int32> 128))) (cast<int32> ?x5))) (cast<int32> 1) (cast<int32> 128))))) (cast<int8x128> (call[x128] (cast<int8> 5))))))))) (cast<int32x128> (call[widening_add] (cast<int16x128> (call[shift_left] (cast<int16x128> (call[int16x128] (cast<uint8x128> (get ?#20 (cast<int32x128> (call[ramp] (cast<int32> (+ (cast<int32> (* (cast<int32> ?x4) (cast<int32> 128))) (cast<int32> ?x6))) (cast<int32> 1) (cast<int32> 128))))))) (cast<uint16x128> (call[x128] (cast<uint16> 3))))) (cast<int16x128> (call[widening_mul] (cast<uint8x128> (get ?#21 (cast<int32x128> (call[ramp] (cast<int32> (+ (cast<int32> (* (cast<int32> ?x4) (cast<int32> 128))) (cast<int32> ?x7))) (cast<int32> 1) (cast<int32> 128))))) (cast<int8x128> (call[x128] (cast<int8> 11))))))))) (cast<int32x128> (call[widening_add] (cast<int16x128> (call[widening_mul] (cast<uint8x128> (get ?#22 (cast<int32x128> (call[ramp] (cast<int32> (+ (cast<int32> (* (cast<int32> ?x8) (cast<int32> 128))) (cast<int32> ?x9))) (cast<int32> 1) (cast<int32> 128))))) (cast<int8x128> (call[x128] (cast<int8> 6))))) (cast<int16x128> (call[widening_mul] (cast<uint8x128> (get ?#23 (cast<int32x128> (call[ramp] (cast<int32> (+ (cast<int32> (* (cast<int32> ?x8) (cast<int32> 128))) (cast<int32> ?x10))) (cast<int32> 1) (cast<int32> 128))))) (cast<int8x128> (call[x128] (cast<int8> 9))))))))",
    ),
    (
        63,
        3,
        "(* (cast<int32> (+ (cast<int32> ?x0) (cast<int32> (neg (cast<int32> 1))))) (cast<int32> ?#1))",
    ),
    (
        32,
        4,
        "(+ (cast<int32> ?x0) (cast<int32> 5))",
    ),
    (
        55,
        4,
        "(- (cast<int32> ?x0) (cast<int32> (call[max] (cast<int32> (call[min] (cast<int32> (* (cast<int32> ?x1) (cast<int32> 4))) (cast<int32> 1))) (cast<int32> (neg (cast<int32> 5))))))",
    ),
    (
        61,
        4,
        "(* (cast<int32> (+ (cast<int32> ?x0) (cast<int32> 1))) (cast<int32> (+ (cast<int32> ?#1) (cast<int32> 128))))",
    ),
    (
        66,
        5,
        "(* (cast<int32> ?x0) (cast<int32> 4))",
    ),
    (
        67,
        5,
        "(* (cast<int32> ?x0) (cast<int32> 128))",
    ),
    (
        68,
        5,
        "(* (cast<int32> ?x0) (cast<int32> ?x1))",
    ),
    (
        23,
        6,
        "(+ (cast<int32> (* (cast<int32> ?x0) (cast<int32> 128))) (cast<int32> ?x1))",
    ),
    (
        64,
        6,
        "(* (cast<int32> (+ (cast<int32> ?x0) (cast<int32> ?x1))) (cast<int32> 4))",
    ),
    (
        24,
        8,
        "(+ (cast<int32> (* (cast<int32> ?x0) (cast<int32> ?x1))) (cast<int32> ?x2))",
    ),
    (
        38,
        12,
        "(+ (cast<int32> ?x0) (cast<int32> ?x1))",
    ),
]
